src/target/riscv/riscv.c

```diff
@@ -852,24 +862,28 @@ static int maybe_add_trigger_t2_t6(struct target *target,
 		struct trigger_request_info lt_1 = {
 			.tdata1 = fields.common | fields.size.any | fields.chain.enable |
 				fields.match.lt,
-			.tdata2 = trigger->address,
+			.tdata2 = trigger->address + trigger->length,
 			.tdata1_ignore_mask = fields.tdata1_ignore_mask
 		};
 		struct trigger_request_info ge_2 = {
 			.tdata1 = fields.common | fields.size.any | fields.chain.disable |
 				fields.match.ge,
-			.tdata2 = trigger->address + trigger->length,
+			.tdata2 = trigger->address,
 			.tdata1_ignore_mask = fields.tdata1_ignore_mask
 		};
 		ret = try_setup_chained_match_triggers(target, trigger, lt_1, ge_2);
 		if (ret != ERROR_TARGET_RESOURCE_NOT_AVAILABLE)
 			return ret;
 	}
 	LOG_TARGET_DEBUG(target, "trying to setup equality match trigger");
+	if (trigger->length > 1)
+		LOG_TARGET_WARNING(target, "Trigger will match accesses at address 0x%"
+				TARGET_PRIxADDR ", but may not match accesses at addresses in the inclusive range from 0x%"
+				TARGET_PRIxADDR " to 0x%" TARGET_PRIxADDR ".", trigger->address,
+				trigger->address + 1, trigger->address + trigger->length - 1);
```

Comment:
I originally misread this check. I thought we were talking about triggers at 0x101 not firing when an access is made to 0x100. But this is the opposite, which is way more common.

I don't think we can print out a warning every time a trigger with length>1 is set. Users will see this message every time they resume a target when such breakpoints are set, because then gdb clears all hardware berakpoints, steps, and sets them all again.

I'm inclined to leave the warning out altogether. If you do think it's important, then please ensure it's displayed only once per session.

Reply:
It may be rare among platforms, but for platforms where this case is hit, it will be hit all the time. I would prefer the warning pop up once, saying something like "on this platform ..." instead of every time the trigger is set (which will be often because gdb clears/sets them for every resume).

Reply:
Addressed.